INSTALL.md

```diff
@@ -1,76 +1,221 @@
 
-## Which version to use?
+If you are using `stack` or `nix` please make sure to add the latest
+version from Hackage to your tool configuration.  -->
 
-* [Latest stable release](https://hackage.haskell.org/package/streamly) is available on
-  Hackage.
-* [Latest development version](https://github.com/composewell/streamly) is
-  available on github.
+No prior knowledge of either Haskell or of `cabal` is needed.  We do
+however assume that you are using a command-line shell on a POSIX
+operating system.  If you are running Windows&trade; then you may need
+to run a command-line shell under `msys` and some of the commands
+below may also need to be changed in small ways to make them work.
 
-If you are using `stack` or `nix` please make sure to add the latest version
-from Hackage to your tool configuration.
+**Note**: If you are new to Haskell you may find the [Haskell Getting
+Started
+Guide](https://github.com/composewell/haskell-dev/blob/master/getting-started.rst)
+to be useful.
 
-## Recommended build options
+If you know your way around Haskell, and have an up to date toolchain
+already installed, then you can jump straight to the section titled
+"[_Prepare Your Build Directory_](#user-content-prepare-your-build-directory)".
 
-See [recommended compilation options here](docs/Build.md).
+## Getting started with `streamly` using `cabal`
 
-## Using with `cabal`
+To get started, you will need a recent `cabal` binary to be installed
+on your system.
 
-Make sure you have `cabal` version 3.0 or later installed and you have `ghc`
-available in your PATH. Refresh your package list:
+### Installing `cabal`
+
+A pre-packaged `cabal` binary may already be available for your
+operating system.
+
+For example, on Debian GNU/Linux, you could install `cabal` using::
+
+```
+$ sudo apt install cabal-install
+```
+
+### Installing `ghc`
+
+You would also need to install the Glasgow Haskell Compiler `ghc`.  In
+most operating systems, using the system's package manager to install
+`cabal` would also bring in a version of `ghc` as a dependency.
+
+Please verify that `ghc` is present:
+
+```
+$ ghc --version
+The Glorious Glasgow Haskell Compilation System, version 8.0.2
+```
+
+If `ghc` isn't present, please install it using your operating
+system's package manager. For example, on Debian GNU/Linux, use:
+
+```
+$ sudo apt install ghc
+```
+
+### Update `cabal` if necessary
+
+Many operating systems ship an old version of `cabal`; in order to use
+`streamly` you should use `cabal` version 3.0 (or later).
+
+You can check the installed version of `cabal` by using its
+`--version` flag:
+
+```
+$ which cabal
+/usr/bin/cabal
+$ /usr/bin/cabal --version
+cabal-install version 1.24.0.2
+compiled using version 1.24.2.0 of the Cabal library
+```
+
+In this example, the installed `cabal` binary is too old for our
+needs. It would need to be upgraded.
+
+You can ask `cabal` to upgrade itself:
+
+```
+$ cabal install Cabal cabal-install
```

Comment:
This is a bit heavy upgrade method, may take a long time. Also, the newer `cabal install` requires `--installdir` option.  This is another reason to prefer `ghcup`. 

Also, I am wondering why we need `Cabal` here, won't just `cabal install cabal-install` work?

Reply:
That is what <https://www.haskell.org/cabal/> recommends.  I've added a link to that page.

Reply:
> This is a bit heavy upgrade method, may take a long time.

It does take a while, but it rebuilds `cabal` from source. If the local Haskell tool chain has not been compromised we should end up with a binary that reflects the source code for `cabal`.

> This is another reason to prefer ghcup.

Unfortunately `ghcup` seems to require us to download and run a binary of unknown provenance.